File: src/ui/tutorial.cpp
#include <saucer/smartview.hpp>
#include "tutorial.h"

#include "../core/scheduler.h"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <format>

// -----------------------------------------------------------------------
// Registries / step data
// -----------------------------------------------------------------------

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wglobal-constructors"
std::vector<Tutorial::Step>			   Tutorial::_steps;
std::vector<BaseElement::TutorialStep> Tutorial::_raw_steps;
saucer::smartview*					   Tutorial::_view;
ui::dom::Element					   Tutorial::_overlay;
ui::dom::Element					   Tutorial::_spotlight;
std::vector<ui::dom::Element>		   Tutorial::_dimmers;
ui::dom::Element					   Tutorial::_panel;
ui::dom::Element					   Tutorial::_panel_title;
ui::dom::Element					   Tutorial::_panel_desc;
ui::dom::Element					   Tutorial::_panel_counter;
ui::dom::Element					   Tutorial::_btn_prev;
ui::dom::Element					   Tutorial::_btn_next;
ui::dom::Element					   Tutorial::_btn_skip;
int									   Tutorial::_index;
bool								   Tutorial::_active;
int									   Tutorial::_gen;
#pragma clang diagnostic pop

void Tutorial::initializeAll(saucer::smartview* view)
{
	_view = view;
	onDomReady();
}

void Tutorial::release()
{
	_active = false;
	++_gen;	   // kill hanging background positioning tasks
	_steps.clear();
	_raw_steps.clear();
	_view = nullptr;
}

void Tutorial::onDomReady()
{
	if (!_view)
		return;

	buildStepList();

	// "Start the tour" button on the #tutorial tab.
	auto container = ui::dom::querySelector("#tutorial_tour_start");
	if (!container.valid())
		return;

	auto btn = ui::dom::create("div");
	btn.addClass("button").show();
	container.append(btn);

	auto inner = ui::dom::create("button");
	inner.text(Localization::Str{ "str_tutorial_tour_start_button" }());
	btn.append(inner);

	btn.on(
		ui::dom::Event::Click,
		[](std::string, js::Value) -> bool
		{
			startTour();
			return false;	 // stay registered
		},
		"tutorial_tour_start"
	);
}

// -----------------------------------------------------------------------
// Step list construction
// -----------------------------------------------------------------------

void Tutorial::buildStepList()
{
	_steps.clear();
	_raw_steps = BaseElement::tutorialSteps();

	// Hand-authored steps: intro + service block (matching the old JS).
	_steps.push_back(Step{ "__intro__", "intro", "str_tour_intro_title", "str_tour_intro_description", "#tutorial", 0, true });
	_steps.push_back(Step{ "#zapret .service", "block", "str_tour_services_title", "str_tour_services_description", "", 5, false });

	for (const auto& raw : _raw_steps)
		_steps.push_back(Step{ raw.name, raw.type, raw.title, raw.desc, "", raw.priority, false });

	std::stable_sort(_steps.begin(), _steps.end(), [](const Step& a, const Step& b) { return a.priority < b.priority; });
}

ui::dom::Element Tutorial::targetFor(const Step& step)
{
	if (step.type == "intro")
		return {};

	if (step.type == "block")
		return ui::dom::querySelector(step.name);

	return BaseElement::element(step.name);
}

// -----------------------------------------------------------------------
// Overlay / steps
// -----------------------------------------------------------------------

void Tutorial::startTour()
{
	if (_active || !_view || _steps.empty())
		return;

	_active = true;
	_index	= 0;

	_overlay = ui::dom::create("div");
	_overlay.id("tour_overlay").addClass("tour_overlay");
	ui::dom::body().append(_overlay);

	_spotlight = ui::dom::create("div");
	_spotlight.id("tour_spotlight").addClass("tour_spotlight");
	_overlay.append(_spotlight);

	_dimmers.clear();

	for (const char* side : { "top", "bottom", "left", "right" })
	{
		auto dimmer = ui::dom::create("div");
		dimmer.addClass("tour_dimmer").addClass(std::string("tour_dimmer_") + side);
		_overlay.append(dimmer);
		_dimmers.push_back(dimmer);
	}

	_panel = ui::dom::create("div");
	_panel.id("tour_panel").addClass("tour_panel");
	_overlay.append(_panel);

	_panel_title = ui::dom::create("h3");
	_panel_title.addClass("tour_panel_title");
	_panel.append(_panel_title);

	_panel_desc = ui::dom::create("p");
	_panel_desc.addClass("tour_panel_desc");
	_panel.append(_panel_desc);

	_panel_counter = ui::dom::create("div");
	_panel_counter.addClass("tour_panel_counter");
	_panel.append(_panel_counter);

	auto buttons = ui::dom::create("div");
	buttons.addClass("tour_panel_buttons");
	_panel.append(buttons);

	_btn_prev = ui::dom::create("button");
	_btn_prev.addClass("tour_btn").addClass("tour_btn_prev");
	_btn_prev.on(
		ui::dom::Event::Click,
		[](std::string, js::Value) -> bool
		{
			if (_index > 0)
			{
				_index--;
				showStep();
			}
			return false;
		},
		"tutorial_tour_prev"
	);
	buttons.append(_btn_prev);

	_btn_skip = ui::dom::create("button");
	_btn_skip.addClass("tour_btn").addClass("tour_btn_skip");
	_btn_skip.on(
		ui::dom::Event::Click,
		[](std::string, js::Value) -> bool
		{
			endTour();
			return false;
		
		},
		"tutorial_tour_skip"
	);
	buttons.append(_btn_skip);

	_btn_next = ui::dom::create("button");
	_btn_next.addClass("tour_btn").addClass("tour_btn_next");
	_btn_next.on(
		ui::dom::Event::Click,
		[](std::string, js::Value) -> bool
		{
			if (_index >= static_cast<int>(_steps.size()) - 1)
				endTour();
			else
			{
				_index++;
				showStep();
			}

			return false;
		
		},
		"tutorial_tour_next"
	);
	buttons.append(_btn_next);

	showStep();
}

void Tutorial::endTour()
{
	if (!_active)
		return;

	_active = false;
	++_gen;	   // kill hanging background positioning tasks
	lockScroll(false);
	// Smooth finale: drop the show classes (CSS transition 0.25s), remove
	// the node after the fade — an instant remove() toggled the dimmers.
	// The old overlay is captured by copy: a new tour within 250ms
	// builds its own.
	if (_spotlight.valid())
		_spotlight.removeClass("tour_spotlight_show");

	for (auto& dim : _dimmers)
		dim.removeClass("tour_dimmer_show");

	if (_panel.valid())
		_panel.addClass("tour_panel_hidden");

	auto old_overlay = _overlay;
	using namespace std::chrono_literals;
	Scheduler::get().after(250ms, [old_overlay]() mutable { old_overlay.remove(); });
	_overlay   = {};
	_spotlight = {};
	_dimmers.clear();
	_panel = {};
}

void Tutorial::showStep()
{
	if (!_active || _steps.empty())
		return;

	const int  gen	= ++_gen;
	const Step step = _steps[static_cast<std::size_t>(_index)];

	updatePanel();
	resetOverlayVisual(gen);

	auto target = targetFor(step);
	if (step.type == "intro" || !target.valid())
		scheduleCenter(gen);
	else
		scheduleLayout(gen, step, target);
}

void Tutorial::lockScroll(bool on)
{
	// The tour is modal: freeze the scroll so the measured positions
	// cannot drift under the fixed overlay.
	ui::dom::querySelector("html").style("overflow", on ? "hidden" : "");
	ui::dom::body().style("overflow", on ? "hidden" : "");
}

std::string Tutorial::resolveTab(const Step& step, const ui::dom::Element& target)
{
	if (!step.tab.empty())
		return step.tab;

	if (target.valid())
	{
		auto art = target.closest("article");
		if (art.valid())
		{
			const auto id = art.getAttr("id");
			if (!id.empty())
				return "#" + id;
		}
	}

	return "#tutorial";
}

void Tutorial::switchTab(std::string_view tab)
{
	if (tab.empty())
		return;
	// Tab switching and URL cosmetics live in menu.js (preventDefault +
	// pushState); here only a click on the nav link.
	ui::dom::querySelector(std::format(".nav a[href=\"{}\"]", tab)).click();
}

void Tutorial::resetOverlayVisual(int gen)
{
	// Dimmers fade via transition (only the class is removed), display:none —
	// after the fade (250ms = CSS transition). The deferral is gen-guarded:
	// a new step within that time already re-shows the dimmers, so a foreign
	// hide must not fight it.
	for (auto& dim : _dimmers)
		dim.removeClass("tour_dimmer_show");

	auto dims = _dimmers;
	using namespace std::chrono_literals;
	Scheduler::get().after(
		250ms,
		[gen, dims]() mutable
		{
			if (stale(gen))
				return;

			for (auto& dim : dims)
				dim.style("display", "none");
		}
	);

	if (_spotlight.valid())
		_spotlight.removeClass("tour_spotlight_show");

	if (_panel.valid())
		_panel.addClass("tour_panel_hidden");
}

bool Tutorial::stale(int gen)
{
	return gen != _gen || !_active;
}

void Tutorial::scheduleCenter(int gen)
{
	// Intro: give the tab a moment to paint (300ms as in the old shim),
	// then center the panel.
	using namespace std::chrono_literals;
	Scheduler::get().after(
		300ms,
		[gen]
		{
			if (stale(gen))
				return;

			centerNow();
		}
	);
}

void Tutorial::scheduleLayout(int gen, Step step, ui::dom::Element target)
{
	// Element step (timings as in the old shim): wait for the tab render,
	// scroll the target to center, and measure only after the smooth scroll.
	using namespace std::chrono_literals;
	Scheduler::get().after(
		900ms,
		[gen, step = std::move(step), target]() mutable
		{
			if (stale(gen))
				return;

			switchTab(resolveTab(step, target));
			target.scrollIntoView(true);

			Scheduler::get().after(
				450ms,
				[gen, target]
				{
					if (stale(gen))
						return;

					if (auto r = target.rect())
						layoutNow(*r);
					else
						centerNow();	// node gone (query into null) — center like intro
				}
			);
		}
	);
}

void Tutorial::layoutNow(const ui::dom::Rect& target_rect)
{
	using namespace ui::dom::tour;

	const ui::dom::Size vp = ui::dom::Element::viewport();
	const auto			ps = _panel.offsetSize();
	if (vp.w <= 0 || vp.h <= 0 || !ps || ps->w <= 0 || ps->h <= 0)
		return;

	// Integral edges: getBoundingClientRect after a smooth scroll yields
	// fractions, and the browser rounds each box independently — 1px gaps/
	// overlaps opened on the shared dimmer/spotlight edges (step-dependent).
	auto box			= padded(target_rect);
	box.left			= std::round(std::max(0.0, box.left));
	box.top				= std::round(std::max(0.0, box.top));
	const double right	= std::round(std::min(vp.w, box.right()));
	const double bottom = std::round(std::min(vp.h, box.bottom()));

	_spotlight.setBox(box.left, box.top, right - box.left, bottom - box.top);
	_spotlight.addClass("tour_spotlight_show");

	if (_dimmers.size() == 4)
	{
		// cssText sets display:block, the show class kicks off the fade-in
		// (pair to resetOverlayVisual, where the class is removed for fade-out).
		const auto css = dimmerCss(vp.w, vp.h, box.left, box.top, right, bottom);
		_dimmers[0].cssText(css.top).addClass("tour_dimmer_show");
		_dimmers[1].cssText(css.bottom).addClass("tour_dimmer_show");
		_dimmers[2].cssText(css.left).addClass("tour_dimmer_show");
		_dimmers[3].cssText(css.right).addClass("tour_dimmer_show");
	}

	const auto pos = placePanel(target_rect, vp, *ps);
	_panel.moveTo(pos.left, pos.top);
	_panel.removeClass("tour_panel_hidden");
}

void Tutorial::centerNow()
{
	const ui::dom::Size vp = ui::dom::Element::viewport();
	const auto			ps = _panel.offsetSize();
	if (vp.w <= 0 || vp.h <= 0 || !ps || ps->w <= 0 || ps->h <= 0)
		return;

	const auto pos = ui::dom::tour::centerPanel(vp, *ps);
	_panel.moveTo(pos.left, pos.top);
	_panel.removeClass("tour_panel_hidden");
}

void Tutorial::updatePanel()
{
	const auto& step = _steps[static_cast<std::size_t>(_index)];

	_panel_title.text(Localization::Str{ step.title }());
	_panel_desc.text(Localization::Str{ step.desc }());

	const auto step_text = Localization::Str{ "str_tour_step" }();
	const auto counter	 = utils::format(step_text, _index + 1, _steps.size());
	_panel_counter.text(counter);

	_btn_prev.style("visibility", _index == 0 ? "hidden" : "visible");
	_btn_prev.text(Localization::Str{ "str_tour_prev" }());
	_btn_skip.text(Localization::Str{ "str_tour_skip" }());

	const auto is_last = _index >= static_cast<int>(_steps.size()) - 1;
	_btn_next.text(Localization::Str{ is_last ? "str_tour_done" : "str_tour_next" }());
}

File: ui/scripts/tutorial.js
// =========================================================
// Interactive tour of the Unblock interface
// Launched by the «Take a tour» button on the #tutorial tab
// =========================================================

let tour_overlay = null;
let tour_spotlight = null;
let tour_dimmers = [];
let tour_panel = null;
let tour_panel_title = null;
let tour_panel_desc = null;
let tour_panel_counter = null;
let tour_btn_prev = null;
let tour_btn_skip = null;
let tour_btn_next = null;
let tour_steps = [];
let tour_index = 0;
let tour_step_timeout = null;
let tour_registry = [];

async function tourLang(_key, _fallback) {
	try {
		if (RUN_CPP)
			return await saucer.exposed.CPPLangText(_key);
	} catch (error) {}

	return _fallback;
}

function tourNormalizeType(_type) {
	if (_type === "check_box")
		return "checkbox";
	if (_type === "select_list")
		return "select";
	return _type;
}

/** Registers a step of the interactive tutorial.
 *  Can be called from C++ (BaseElement::addTutorialStep) or directly from JS.
 *  @param _name       element name (for a block — a CSS selector).
 *  @param _title_id   localization key of the step title.
 *  @param _desc_id    localization key of the step description.
 *  @param _priority   optional priority (by default — appended to the end).
 *  @param _type       element type: button | checkbox | select | block | intro. */
function registerTutorialStep(_name, _title_id, _desc_id, _priority, _type) {
	const priority = (_priority === undefined || _priority === null) ? Number.MAX_SAFE_INTEGER : Number(_priority);
	tour_registry.push({
		name: _name,
		type: tourNormalizeType(_type),
		title: _title_id,
		desc: _desc_id,
		priority: priority
	});
	return true;
}

// Build the list of steps. Steps whose elements are not found or hidden are skipped.
function tourBuildSteps() {
	const steps = [];

	for (const reg of tour_registry) {
		if (reg.type === "intro") {
			steps.push({ ...reg, tab: "#tutorial", intro: true });
			continue;
		}

		const element = tourGetElement(reg);
		if (!tourIsVisible(element))
			continue;

		const article = element.closest ? element.closest("article") : null;
		const tab = article && article.id ? "#" + article.id : "#tutorial";
		steps.push({ ...reg, tab: tab });
	}

	steps.sort((a, b) => a.priority - b.priority);
	return steps;
}

function tourGetElement(step) {
	if (step.type === "button") {
		return getButton(step.name);
	} else if (step.type === "checkbox") {
		return getCheckBox(step.name);
	} else if (step.type === "select") {
		const select = getSelect(step.name);
		return select ? select.div : undefined;
	} else if (step.type === "block") {
		return document.querySelector(step.name);
	}

	return undefined;
}

function tourIsVisible(element) {
	if (!element)
		return false;

	const rect = element.getBoundingClientRect();
	return rect.width > 0 && rect.height > 0;
}

// ============================ Overlay ============================

function startTutorialTour() {
	if (tour_overlay || !document.body)
		return;

	tour_steps = tourBuildSteps();
	if (tour_steps.length === 0)
		return;

	tour_index = 0;
	tourBuildOverlay();
	tourShowStep();
}

function tourBuildOverlay() {
	tour_overlay = document.createElement("div");
	tour_overlay.id = "tour_overlay";
	tour_overlay.classList.add("tour_overlay");
	document.body.appendChild(tour_overlay);

	tour_spotlight = document.createElement("div");
	tour_spotlight.id = "tour_spotlight";
	tour_spotlight.classList.add("tour_spotlight");
	tour_overlay.appendChild(tour_spotlight);

	tour_dimmers = [];
	for (const side of ["top", "bottom", "left", "right"]) {
		const dimmer = document.createElement("div");
		dimmer.classList.add("tour_dimmer", "tour_dimmer_" + side);
		tour_overlay.appendChild(dimmer);
		tour_dimmers.push(dimmer);
	}

	tour_panel = document.createElement("div");
	tour_panel.id = "tour_panel";
	tour_panel.classList.add("tour_panel");
	tour_overlay.appendChild(tour_panel);

	tour_panel_title = document.createElement("h3");
	tour_panel_title.classList.add("tour_panel_title");
	tour_panel.appendChild(tour_panel_title);

	tour_panel_desc = document.createElement("p");
	tour_panel_desc.classList.add("tour_panel_desc");
	tour_panel.appendChild(tour_panel_desc);

	tour_panel_counter = document.createElement("div");
	tour_panel_counter.classList.add("tour_panel_counter");
	tour_panel.appendChild(tour_panel_counter);

	const buttons = document.createElement("div");
	buttons.classList.add("tour_panel_buttons");
	tour_panel.appendChild(buttons);

	tour_btn_prev = document.createElement("button");
	tour_btn_prev.classList.add("tour_btn", "tour_btn_prev");
	buttons.appendChild(tour_btn_prev);

	tour_btn_skip = document.createElement("button");
	tour_btn_skip.classList.add("tour_btn", "tour_btn_skip");
	buttons.appendChild(tour_btn_skip);

	tour_btn_next = document.createElement("button");
	tour_btn_next.classList.add("tour_btn", "tour_btn_next");
	buttons.appendChild(tour_btn_next);

	tour_btn_prev.addEventListener("click", () => {
		if (tour_index > 0) {
			tour_index--;
			tourShowStep();
		}
	});

	tour_btn_skip.addEventListener("click", tourEnd);

	tour_btn_next.addEventListener("click", () => {
		if (tour_index >= tour_steps.length - 1) {
			tourEnd();
			return;
		}
		tour_index++;
		tourShowStep();
	});
}

function tourEnd() {
	if (tour_step_timeout)
		clearTimeout(tour_step_timeout);

	if (tour_overlay) {
		tour_overlay.remove();
		tour_overlay = null;
		tour_spotlight = null;
		tour_dimmers = [];
		tour_panel = null;
	}
}

// ============================ Step ============================

function tourShowStep() {
	if (!tour_overlay || tour_steps.length === 0)
		return;

	if (tour_step_timeout)
		clearTimeout(tour_step_timeout);

	const step = tour_steps[tour_index];

	tour_spotlight.classList.remove("tour_spotlight_show");
	tourHideDimmers();

	// Hide the panel (fade-out) before switching to the next step
	tour_panel.classList.add("tour_panel_hidden");

	if (step.intro) {
		tourSwitchTab(step.tab);
		tour_step_timeout = setTimeout(() => {
			tourUpdatePanelContent(step).then(() => {
				tourPositionPanel();
				tourShowPanel();
			});
		}, 300);
		return;
	}

	tour_step_timeout = setTimeout(() => {
		tourSwitchTab(step.tab);

		tour_step_timeout = setTimeout(() => {
			const element = tourGetElement(step);
			if (!tourIsVisible(element)) {
				tourNextStep(true);
				return;
			}

			if (element && element.scrollIntoView)
				element.scrollIntoView({ block: "center", behavior: "smooth" });

			tour_step_timeout = setTimeout(() => {
				tourUpdatePanelContent(step).then(() => {
					tourShowSpotlight(element);
					tourPositionPanel(element);
					tourShowPanel();
				});
			}, 450);
		}, 900);
	}, 300);
}

async function tourUpdatePanelContent(step) {
	tour_panel_title.textContent = await tourLang(step.title, step.title);
	tour_panel_desc.textContent = await tourLang(step.desc, step.desc);
	tour_panel_counter.textContent = await tourFormatCounter(tour_index + 1, tour_steps.length);

	tour_btn_prev.style.visibility = tour_index === 0 ? "hidden" : "visible";
	tour_btn_prev.textContent = await tourLang("str_tour_prev", "← Back");

	tour_btn_skip.textContent = await tourLang("str_tour_skip", "Skip");

	const is_last = tour_index >= tour_steps.length - 1;
	tour_btn_next.textContent = await tourLang(is_last ? "str_tour_done" : "str_tour_next", is_last ? "✓ Done" : "Next →");
}

function tourShowPanel() {
	tour_panel.classList.remove("tour_panel_hidden");
}

async function tourFormatCounter(current, total) {
	const text = await tourLang("str_tour_step", "{} of {}");
	return text.replace("{}", String(current)).replace("{}", String(total));
}

function tourSwitchTab(tab) {
	const link = document.querySelector(`.nav a[href="${tab}"]`);
	if (link)
		link.click();
}

function tourHideDimmers() {
	for (const dimmer of tour_dimmers)
		dimmer.style.display = "none";
}

function tourShowSpotlight(element) {
	if (!element)
		return;

	const rect = element.getBoundingClientRect();
	const padding = 8;
	const viewport_width = window.innerWidth;
	const viewport_height = window.innerHeight;

	const left = Math.max(0, rect.left - padding);
	const top = Math.max(0, rect.top - padding);
	const right = Math.min(viewport_width, rect.right + padding);
	const bottom = Math.min(viewport_height, rect.bottom + padding);

	tour_spotlight.style.left = left + "px";
	tour_spotlight.style.top = top + "px";
	tour_spotlight.style.width = (right - left) + "px";
	tour_spotlight.style.height = (bottom - top) + "px";
	tour_spotlight.classList.add("tour_spotlight_show");

	const dimmer_style = ["top", "bottom", "left", "right"];

	for (const side of dimmer_style)
		tourDimmer(side).style.display = "block";

	tourDimmer("top").style.left = "0px";
	tourDimmer("top").style.top = "0px";
	tourDimmer("top").style.width = "100%";
	tourDimmer("top").style.height = top + "px";

	tourDimmer("bottom").style.left = "0px";
	tourDimmer("bottom").style.top = bottom + "px";
	tourDimmer("bottom").style.width = "100%";
	tourDimmer("bottom").style.height = Math.max(0, viewport_height - bottom) + "px";

	tourDimmer("left").style.left = "0px";
	tourDimmer("left").style.top = top + "px";
	tourDimmer("left").style.width = left + "px";
	tourDimmer("left").style.height = (bottom - top) + "px";

	tourDimmer("right").style.left = right + "px";
	tourDimmer("right").style.top = top + "px";
	tourDimmer("right").style.width = Math.max(0, viewport_width - right) + "px";
	tourDimmer("right").style.height = (bottom - top) + "px";
}

function tourDimmer(side) {
	return tour_dimmers[["top", "bottom", "left", "right"].indexOf(side)];
}

function tourPositionPanel(element) {
	const panel_width = tour_panel.offsetWidth;
	const panel_height = tour_panel.offsetHeight;
	const viewport_width = window.innerWidth;
	const viewport_height = window.innerHeight;

	if (!element) {
		tour_panel.style.left = Math.max(16, (viewport_width - panel_width) / 2) + "px";
		tour_panel.style.top = Math.max(16, (viewport_height - panel_height) / 2) + "px";
		return;
	}

	const rect = element.getBoundingClientRect();
	const gap = 24;
	const margin = 16;

	const space_left = rect.left - margin - gap;
	const space_right = viewport_width - rect.right - margin - gap;

	if (space_left >= panel_width || space_right >= panel_width) {
		const on_right_side = rect.left + rect.width / 2 > viewport_width / 2;

		let left;
		if (on_right_side) {
			left = rect.left - panel_width - gap;
			if (left < margin)
				left = rect.right + gap;
			left = Math.min(left, viewport_width - panel_width - margin);
		} else {
			left = rect.right + gap;
			if (left + panel_width > viewport_width - margin)
				left = rect.left - panel_width - gap;
			left = Math.max(margin, left);
		}

		let top = rect.top + rect.height / 2 - panel_height / 2;
		top = Math.max(margin, Math.min(top, viewport_height - panel_height - margin));

		tour_panel.style.left = left + "px";
		tour_panel.style.top = top + "px";
		return;
	}

	const space_top = rect.top - margin - gap;
	const space_bottom = viewport_height - rect.bottom - margin - gap;

	const place_bottom = space_bottom >= panel_height && space_bottom >= space_top;

	let top;
	if (place_bottom)
		top = rect.bottom + gap;
	else
		top = rect.top - panel_height - gap;

	top = Math.max(margin, Math.min(top, viewport_height - panel_height - margin));

	let left = (viewport_width - panel_width) / 2;
	left = Math.max(margin, Math.min(left, viewport_width - panel_width - margin));

	tour_panel.style.left = left + "px";
	tour_panel.style.top = top + "px";
}

function tourNextStep(force) {
	if (force && tour_index < tour_steps.length - 1) {
		tour_index++;
		tourShowStep();
	}
}

// ============================ Init ============================

async function tutorialTourInit() {
	registerTutorialStep("__intro__", "str_tour_intro_title", "str_tour_intro_description", 0, "intro");
	registerTutorialStep("#zapret .service", "str_tour_services_title", "str_tour_services_description", 5, "block");

	const container = document.getElementById("tutorial_tour_start");
	if (!container)
		return;

	createButton("#tutorial_tour_start", "tutorial_tour_start", await tourLang("str_tutorial_tour_start_button", "▶ Start tutorial"));

	const button = getButton("tutorial_tour_start");
	if (button) {
		const native_button = button.firstChild;
		if (native_button)
			native_button.addEventListener("click", startTutorialTour);
	}
}

if (document.readyState === "loading")
	document.addEventListener("DOMContentLoaded", () => { tutorialTourInit(); });
else
	tutorialTourInit();
